robocop/checkers/naming.py

```diff
@@ -378,3 +380,70 @@ def visit_KeywordCall(self, node):  # noqa
             token = node.data_tokens[1]
             if token.type == Token.ARGUMENT and not token.value.isupper():
                 self.report("non-local-variables-should-be-uppercase", node=node, col=token.col_offset + 1)
+
+
+class SimilarVariableChecker(VisitorChecker):
+    """ Checker for finding same variables with similar names. """
+    rules = {
+        "0316": (
+            "possible-variable-overwriting",
+            "Variable '%s' may overwrite similar variable inside '%s' %s. "
+            "Note that variables are case-insensitive, and also spaces and underscores are ignored.",
+            RuleSeverity.INFO
+        )
+    }
+
+    def visit_Keyword(self, node):  # noqa
+        self.visit_vars_and_find_similar(node)
+        self.generic_visit(node)
+
+    def visit_TestCase(self, node):  # noqa
+        self.visit_vars_and_find_similar(node)
+        self.generic_visit(node)
+
+    def visit_vars_and_find_similar(self, node):
+        """
+        Creates a dictionary `variables` with normalized variable name as a key
+        and ads a list of all detected variations of this variable in the node as a value,
+        then it checks if similar variable was found.
+        """
+        variables = defaultdict(set)
+
+        for child in node.body:
+            # read arguments from Test Case or Keyword
+            if isinstance(child, Arguments):
+                for token in child.get_tokens(Token.ARGUMENT):
+                    variables[normalize_robot_var_name(token.value)].add(token.value)
+
+            # check KEYWORD CALLS
+            elif hasattr(child, 'keyword'):
+                tokens = child.get_tokens(Token.ASSIGN)
+                self.find_similar_variables(tokens, variables, node)
+
+            # check FOR LOOPS
+            elif isinstance(child, For):
+                # look for nested for loops
+                while child.body and isinstance(child.body[0], For):
+                    for var in child.variables:
+                        variables[normalize_robot_var_name(var)].add(var)
+                    child = child.body[0]
```

Comment:
What if its nested for loop and then two adjacent for loops? I think it would be better to create variable default dict globally, reset it before entering keyword / test case and then gather and check variables using visitors. It will take care of nested structures for you.

Reply:
Yeah, you're right. I tried using visitors but I again failed with nested structures so I did this. Now I see, that I should stick with visitors. I'm tired of this issue already 😛